test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc

```diff
@@ -480,6 +639,173 @@ cluster: fake_cluster
+
+  test_sessions_[0].recvDataFromUpstream("world");
+  checkTransferStats(5 /*rx_bytes*/, 1 /*rx_datagrams*/, 5 /*tx_bytes*/, 1 /*tx_datagrams*/);
+}
+
+// Make sure socket option is not set correctly if use_original_src_ip is not mentioned.
+TEST_F(UdpProxyFilterTestIpv4Ipv6, NoSocketOptionForNoUseOriginalSrcIp) {
+  if (!isTransparentSocketOptionsSupported()) {
+    // The option is not supported on this platform. Just skip the test.
+    GTEST_SKIP();
+  }
+
+  InSequence s;
+
+  setup(R"EOF(
+stat_prefix: foo
+cluster: fake_cluster
+)EOF");
+
+  expectSessionCreate(upstream_address_v6_);
+  test_sessions_[0].expectUpstreamWrite("hello");
+  recvDataFromDownstream("[2001:db8:85a3::9a2e:370:7334]:1000", "[2001:db8:85a3::9a2e:370:7335]:80",
+                         "hello");
+
+  checkSocketOptions(test_sessions_[0], ENVOY_SOCKET_IP_TRANSPARENT, 0,
+                     ENVOY_SOCKET_IPV6_TRANSPARENT, 0);
+  EXPECT_EQ(1, config_->stats().downstream_sess_total_.value());
+  EXPECT_EQ(1, config_->stats().downstream_sess_active_.value());
+  checkTransferStats(5 /*rx_bytes*/, 1 /*rx_datagrams*/, 0 /*tx_bytes*/, 0 /*tx_datagrams*/);
+
+  test_sessions_[0].recvDataFromUpstream("world");
+  checkTransferStats(5 /*rx_bytes*/, 1 /*rx_datagrams*/, 5 /*tx_bytes*/, 1 /*tx_datagrams*/);
+}
+
+// Make sure socket option is not set correctly if the platform does not support the socket option.
+TEST_F(UdpProxyFilterTest, FallbackToNoOrgSrcIpForNoPlatformSupport) {
+  InSequence s;
+
+  EXPECT_CALL(os_sys_calls_, supportsIpTransparent());
+
+  setup(R"EOF(
+stat_prefix: foo
+cluster: fake_cluster
+use_original_src_ip: true
+)EOF");
+
+  expectSessionCreate(upstream_address_);
+  test_sessions_[0].expectUpstreamWriteOriginalSrcIp(true);
+  test_sessions_[0].expectUpstreamWrite("hello", 0);
+  recvDataFromDownstream(peer_address_->asString(), "10.0.0.2:80", "hello");
+
+  if (ENVOY_SOCKET_IP_TRANSPARENT.hasValue() && ENVOY_SOCKET_IPV6_TRANSPARENT.hasValue()) {
+    checkSocketOptions(test_sessions_[0], ENVOY_SOCKET_IP_TRANSPARENT, 0,
+                       ENVOY_SOCKET_IPV6_TRANSPARENT, 0);
+  }
+  EXPECT_EQ(1, config_->stats().downstream_sess_total_.value());
+  EXPECT_EQ(1, config_->stats().downstream_sess_active_.value());
+  checkTransferStats(5 /*rx_bytes*/, 1 /*rx_datagrams*/, 0 /*tx_bytes*/, 0 /*tx_datagrams*/);
+
+  test_sessions_[0].recvDataFromUpstream("world");
+  checkTransferStats(5 /*rx_bytes*/, 1 /*rx_datagrams*/, 5 /*tx_bytes*/, 1 /*tx_datagrams*/);
+}
+
+// Make sure exit when use the use_original_src_ip but platform does not support ip transparent
+// option.
+TEST_F(UdpProxyFilterTest, ExitIpTransparentNoPlatformSupport) {
+  InSequence s;
+
+  auto config = R"EOF(
+stat_prefix: foo
+cluster: fake_cluster
+use_original_src_ip: true
+)EOF";
+
+  EXPECT_CALL(os_sys_calls_, supportsIpTransparent()).WillOnce(Return(false));
+  EXPECT_THROW_WITH_REGEX(setup(config, true), EnvoyException, "The platform does not support .*");
```

Comment:
Can you add more text here to make sure we are catching the right exception?

Reply:
Sure.